Load variation font faces with theme relative urls using backend provided full urls (_links)

matiasbenedetto opened the pull request:
## What?
Use backend provided full urls to load font assets with theme relative urls.

## Why?
To load font faces dynamically in the browser, for example, when they change as a result of applying a theme style variation.
Fixes: https://github.com/WordPress/gutenberg/issues/59965

## How?
Uses the links provided by the global styles rest API to load the font assets using the full URLs instead of the theme-relative ones that don't work on the client.


## Testing Instructions
Try to load a style variation and check that the fonts referenced in those style variations load as expected. See the screencast below as example.

**I created a theme to test this quickly and visually:**
[font-variants.zip](https://github.com/user-attachments/files/17311496/font-variants.zip)

| Before |  After |
| --- | --- |
| The fonts defined in the variants are not loaded, so the fallback system fonts are rendered by the browser. | The fonts are loaded so the expected font assets are rendered. |



### Before:
[Screencast from 09-10-24 13:25:05.webm](https://github.com/user-attachments/assets/3ec7d144-8654-4228-a7d2-b297c901a349)

### After:
[Screencast from 09-10-24 13:23:40.webm](https://github.com/user-attachments/assets/7fa59bce-5360-477d-bde5-09368d02f6ef)



## Screenshots or screencast <!-- if applicable -->
Testing with TT4. It is the same as with the custom  theme I created but the typographies are not so different visually so the changes are harder to see.




https://github.com/user-attachments/assets/adb2c5c1-bf84-40da-933a-e6a9513936b4

matiasbenedetto committed:
add editor font face resolver component

matiasbenedetto committed:
remove code not needed

matiasbenedetto committed:
default value when font families aren't defined

jffng commented:
I tested with TT5 after moving all the font family definitions to their respective style variation / typography presets, and it is working as expected. I haven't reviewed the code yet. 

https://github.com/user-attachments/assets/4ccf11e9-7366-43fe-af85-2a353898b050

jffng reviewed (commented):
I prepared a PR for TT5 to test, which moves all the font family definitions to their respective style variation / typography presets, and it is working as expected. I haven't reviewed the code yet. 

https://github.com/user-attachments/assets/4ccf11e9-7366-43fe-af85-2a353898b050

mikachan commented on packages/block-editor/src/components/editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,63 @@
+	const { currentTheme } = useSelect( ( select ) => {
+		return {
+			currentTheme:
+				// Using 'core' as string to avoid circular dependency importing from @wordpress/core-data
+				select( 'core' ).getCurrentTheme(), // eslint-disable-line @wordpress/data-no-store-string-literals
```
```suggestion
				// Disable Reason: Using 'core' as string to avoid circular dependency importing from @wordpress/core-data.
				// eslint-disable-next-line @wordpress/data-no-store-string-literals
				select( 'core' ).getCurrentTheme(),
```

This is just a slight formatting suggestion based on how I see this used elsewhere in the codebase.

mikachan commented on packages/block-editor/src/components/editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,63 @@
+		};
+	}, [] );
+
+	//Get the fonts from merged theme.json settings.fontFamilies
```
```suggestion
	// Get the fonts from merged theme.json settings.fontFamilies.
```

mikachan commented on packages/block-editor/src/components/editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,63 @@
+
+	useEffect( () => {
+		fontFaces.forEach( loadFontFaceAsset );
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
It looks like `react-hooks/exhaustive-deps` has been disabled because it requires moving `loadFontFaceAsset` into a callback function. Could we try that instead of disabling this rule?

matiasbenedetto replied:
done here: https://github.com/WordPress/gutenberg/pull/65019/commits/0bed01d59389a691988c926e12bb6ba552d8002f

mikachan commented on packages/block-editor/src/components/editor-fonts-resolver/utils.js:
```diff
@@ -0,0 +1,111 @@
+	if ( output.includes( ',' ) ) {
+		output = output
+			.split( ',' )
+			// finds the first item that is not an empty string.
```
```suggestion
			// Finds the first item that is not an empty string.
```

mikachan commented on packages/block-editor/src/components/editor-fonts-resolver/utils.js:
```diff
@@ -0,0 +1,111 @@
+			.find( ( item ) => item.trim() !== '' )
+			.trim();
+	}
+	// removes leading and trailing quotes.
```
```suggestion
	// Removes leading and trailing quotes.
```

mikachan commented on packages/block-editor/src/components/editor-fonts-resolver/utils.js:
```diff
@@ -0,0 +1,111 @@
+
+	if ( typeof source === 'string' ) {
+		dataSource = `url(${ source })`;
+		// eslint-disable-next-line no-undef
```
I'm not seeing any linter errors if I remove this `no-undef` override. Could we remove it?

matiasbenedetto replied:
done here: https://github.com/WordPress/gutenberg/pull/65019/commits/48bfe657541705df1a6961a0c533b9533fd607e4

mikachan reviewed (commented):
This is also working well for me 🎉  I've left some code formatting-related comments.

matiasbenedetto committed:
comments formatting

Co-authored-by: Sarah Norris <1645628+mikachan@users.noreply.github.com>

matiasbenedetto committed:
use callback for loadFontFaceAsset

matiasbenedetto committed:
improve syntax

mikachan commented on packages/block-editor/src/components/block-canvas/index.js:
```diff
@@ -43,6 +44,7 @@ export function ExperimentalBlockCanvas( {
 				__unstableContentRef={ localRef }
 				style={ { height, display: 'flex' } }
 			>
+				<EditorFontsResolver />
```
@youknowriad or @ellatrix since you've edited this file before, could you please sense-check how the fonts resolver is being added here?

mikachan reviewed (commented):
The latest changes are looking good to me, nice work 👏 

I've never edited any code related to the block canvas before; I'm wondering if we should get some more 👀 on this before bringing it in.

justintadlock commented:
This seems to work well in the editor content area and loads up everything as expected.

The two areas where I'm still seeing fallback fonts from the stack are under the Typesets panel:

![image](https://github.com/user-attachments/assets/641004dd-b1ee-47f6-ab22-0d097f3213d0)

And in the Styles sidebar (both the theme and typography style variation sections):

![image](https://github.com/user-attachments/assets/f8f45b2b-a7be-4fd6-bef9-f9ff8681e0e2)

Basically, what the user sees as options is not matching what they're getting in the editor content area.

youknowriad commented on packages/block-editor/src/components/editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,64 @@
+	}, [] );
+
+	// Get the fonts from merged theme.json settings.fontFamilies.
+	const [ fontFamilies = [] ] = useGlobalSetting( 'typography.fontFamilies' );
```
I know it's hard to figure that out on your own but If I'm not wrong `useGlobalSetting` doesn't work consistently in all editors. In fact, we should probably avoid using it in the "block-editor" package entirely because it relies on the presence of a provider that may or may not exist (it doesn't exist in the post editor for instance).

@aaronrobertshaw added a recently a new block editor setting that allows us to access global styles in all editors if I'm not wrong and we should probably be using that.

aaronrobertshaw replied:
Thanks for the ping 👍 

> a new block editor setting that allows us to access global styles in all editors

This was specifically related to the **_styles_** data from Global Styles as the settings values were already included within the block editor settings. 

The Global Styles settings and styles weren't merged as there wasn't consensus on what the shape of those settings should be while maintaining BC. That led to [using a symbol as a private key](https://github.com/WordPress/gutenberg/pull/61556/files#diff-75d4f4d6d3bd43a48a5764edf03d97a190d807658245ff7a8d21cf197222eed3R1) for the Global Styles style data.

Here's the PR adding the style data to the block editor settings: https://github.com/WordPress/gutenberg/pull/61556.

Without looking into the code I'm not sure whether the Global Styles settings under `__experimentalFeatures` in the editor settings are kept in sync with unsaved Global Styles changes in the site editor. It could be possible that aspect needs addressing or perhaps in the short term this could fallback to the editor settings if the Global Styles context isn't available (in post editor).

youknowriad replied:
Good point, I mistakenly thought this is about the styles. 

This PR reminds of the "styles" prop that we pass to the block editor to render the current theme styles. I could suggest a new "type" within the same "styles" prop to load fonts (like we do for svg for instance) but first. I would like to understand more.

What fonts are we trying to resolve here. I'm guessing the fonts that are used in the theme styles right? In which case, I feel like "resolving them" or "rendering them" should be done in the [EditorStyles](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/editor-styles/index.js) component. It's a component that is already used to "resolve or render" SVGs and CSS, it's already rendered in both iframe and non-iframe, so I'm starting to think that it's probably the right place for this change? (Introduce a new "type" of style and render it within that component)

Please correct me if my argumentation or abstraction seem too far off

matiasbenedetto replied:
I'm not sure how to access global styles settings in a component like [EditorStyles](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/editor-styles/index.js) without using `useGlobalSetting`. For this particular implementation I'm interested in getting the 'client side latest  version' of `settings.typography.fontFamilies`.

>  as the settings values were already included within the block editor settings

@aaronrobertshaw could you show me how is this done? I haven't found how to access the fontFamilies path mentioned before.

> What fonts are we trying to resolve here. I'm guessing the fonts that are used in the theme styles right?

Yes. This PR was motivated by that, but implementing this would make us able to remove the manual loading of fonts when a font is installed using the font library, too, so we could simplify a little bit that code.

> I feel like "resolving them" or "rendering them" should be done in the [EditorStyles](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/editor-styles/index.js) component.

@youknowriad I'm not sure about that, EditorStyles gets styles and not settings. At least, as it is today, that component is not getting the data needed to implement the desired functionality.

youknowriad replied:
> I'm not sure about that, EditorStyles gets styles and not settings. At least, as it is today, that component is not getting the data needed to implement the desired functionality.

EditorStyles renders the CSS variables for the color presets for instance, renders the SVG filters used by Duotone. Isn't it the same for fonts here? We're resolving the fonts that are used in the CSS.

It is possible that I'm misunderstanding the issue that this PR solves though and if it's the case please clarify? I'm assuming that resolving the fonts is done in the frontend somehow right? How is it done today?

> I'm not sure how to access global styles settings in a component like [EditorStyles](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/editor-styles/index.js) without using useGlobalSetting. For this particular implementation I'm interested in getting the 'client side latest version' of settings.typography.fontFamilies.

The styles prop received by the `EditorStyles` component can contain things like that

```css
const styles = [
	{
		css: `some css`,
	},
	{
                __unstableType: 'svgs'
                assets: `some sig element`
	},
	{
                __unstableType: 'presets'
                assets: `css with CSS variables generated from the presets`
	},
];
```

This represents all what's necessary for the "styles" to be rendered properly and to work properly.
Would it be out of the question to consider "fonts" at the same level as these and just add

```
{
              __unstableType: 'fonts'
              fonts: `whatever is needed to resolve the fonts`
},
```

Again, it's possible that I'm wrong

youknowriad replied:
Btw, right now that "styles" prop come from the block editor settings array that is passed from the backend to the frontend during the initialization of the editor. So custom CSS files are resolved there for instance and added to that array of `styles`

In the site editor, since we have the ability to update global styles, that "styles" setting (the one that comes from backend to server) is also updated on the fly as we make changes to the global styles Here https://github.com/WordPress/gutenberg/blob/e1ad8c216c0c782bc007e29cf15f82213645f865/packages/edit-site/src/components/global-styles-renderer/index.js#L38

aaronrobertshaw replied:
> could you show me how is this done? I haven't found how to access the fontFamilies path mentioned before.

In the `block-editor` store, under `settings.__experimentalFeatures.typography.fontFamilies`, you should find the data you're chasing in the post editor.

One example of accessing theme.json settings via the block editor store can be found in the [layout block support](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/layout.js?plain=1#L81). Another example (that might change soon) is the [block style variations block support](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/block-style-variation.js?plain=1#L259-L266) which prefers the GlobalStyleContext data, if available, falling back to the block editor store's data.

Hope that helps a little 🤞

matiasbenedetto replied:
Thanks for the insights.
In this commit I removed the use `useGlobalSetting` and replaced it by the use data from block-editor store and move the call to the font faces resolved inside the EditorStyles component: https://github.com/WordPress/gutenberg/pull/65019/commits/c6a32f9d7381e4f93a3c471de7214023f4984edd

youknowriad commented on packages/block-editor/src/components/editor-fonts-resolver/utils.js:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	if ( addTo === 'iframe' || addTo === 'all' ) {
+		const iframeDocument = document.querySelector(
```
Did we consider moving this logic elsewhere. I feel we should avoid using DOM APIs like that, it's really not clear which "editor-canvas" this is targeting. I can see a world where there are two editors iframes on the same page.

If fonts need to be loaded in the iframe, can we instead move the logic to load fonts into the `Iframe` component instead?

We can probably use a memoized selector to fetch the fonts from the global styles config we have in the global styles config key of the block editor settings and load the fonts when they change. I believe we may have logic that get close to this for "styles"...

matiasbenedetto replied:
> If fonts need to be loaded in the iframe, can we instead move the logic to load fonts into the Iframe component instead?

Fonts should be added in both the main document and the iframe's document. If we move that logic to the iframe document, could we add those to the main one with the same code? Should we repeat the font loading in both places? What happen when the editor is not rendering the iframe? (`shouldIframe` is false in a few places across the codebase)

youknowriad replied:
I think in most places for block themes (the ones that supports fonts), we're using the iframe. There are still some case where we are not though: In the post editor, and only if v<3 blocks are registered or no metaboxes and only in Core (Gutenberg plugin forces iframe for the post editor I think in block themes).

We are working on removing these edge cases but yeah in the meantime we can probably have a reusable hook or something and I think `BlockCanvas` is where the switching happens between iframe or not.

matiasbenedetto replied:
In this commit I removed this hardcoded references to the document and 'iframe' references. Now it's using a `ref` to reference the document following the `useDarkThemeBodyClassName` example: https://github.com/WordPress/gutenberg/pull/65019/commits/5a9e2d0cc698aec84edf6e3809f8f379a537325c

matiasbenedetto committed:
Merge branch 'trunk' into add/editor-font-face-resolver

matiasbenedetto committed:
Move EditorFontsResolver inside EditorStyles, use fontFamilies data from block-editor store

matiasbenedetto committed:
use a ref to reference the current document

matiasbenedetto committed:
currentTheme default to empty object

youknowriad commented on packages/block-editor/src/components/editor-styles/index.js:
```diff
@@ -18,53 +18,49 @@ import { useSelect } from '@wordpress/data';
-			if ( ! node ) {
-				return;
-			}
+function useDarkThemeBodyClassName( scope, ref ) {
```
Why did you refactor this from a hook that returns a ref to a hook that receives a ref as an argument?

Returning a ref allow supporting conditional rendering, you can adapt the hook when the "node" changes, it's not something you can do in hooks that receive refs. 

Receiving refs in hooks is a pattern we should avoid (See recent discussion here https://github.com/WordPress/gutenberg/pull/64943#discussion_r1741052415 )

matiasbenedetto replied:
> Why did you refactor this from a hook that returns a ref to a hook that receives a ref as an argument?

Because it's hard to understand the pattern (at least for me :laughing: ), so I thought it wasn't something 100% intentional. Thanks for providing context. Now I see the reasons why it is like that. I reverted the changes to that hook and refactored the new useEditorFontsResolver to follow that pattern: https://github.com/WordPress/gutenberg/pull/65019/commits/463fa952b9cb2170e89398a36625f4195faca6eb

matiasbenedetto committed:
revert changes on useDarkThemeBodyClassName, refactor useEditorFontsResolver follwing useDarkThemeBodyClassName pattern

matiasbenedetto committed:
revert not needed change

mikachan reviewed (commented):
This tests well for me in the Editor; I can see the correct font faces being loaded and applied when switching style variations.

The change to how the fonts resolver is now added to `EditorStyles` looks good and seems more aligned with the other style element used here.

I also believe the feedback in [this comment](https://github.com/WordPress/gutenberg/pull/65019#discussion_r1763221753) from @youknowriad has been addressed, although it'd be great to see if Riad has any further feedback here.

Otherwise, I think this is looking ready to bring in 👏

matiasbenedetto commented:
@justintadlock, could you provide the theme you are testing?

youknowriad commented on packages/block-editor/src/components/use-editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,72 @@
+			currentTheme:
+				// Disable Reason: Using 'core' as string to avoid circular dependency importing from @wordpress/core-data.
+				// eslint-disable-next-line @wordpress/data-no-store-string-literals
+				select( 'core' )?.getCurrentTheme(),
```
This is not allowed in the block-editor package. We shouldn't add an eslint exception. Why do we need the theme object here?

matiasbenedetto replied:
To get the base url of the theme's fonts.

Example: the theme.json data has something like `file:/assets/fonts/one-font.woff2`.
We need to know how to resolve that theme relative font URL in the browser.

youknowriad replied:
We need to find an alternative solution. The block editor is a generic package independent of WordPress and can't make REST API calls.

The alternative I suggested initially is to "resolve" the files in the server and add a new `_unstableType` for the "styles" object received by the editor.

matiasbenedetto replied:
> to "resolve" the files in the server

Do you mean when the `theme.json` data is read from the file / database?

> add a new _unstableType for the "styles" object received by the editor.

Sorry but I don't understand this.

youknowriad replied:
> Do you mean when the theme.json data is read from the file / database?

Yes, basically in this place, in the code base. We read global styles and enqueue "styles" setting in the block editor. We could add a new "style" in that array with a specific `"_unstableType" => "font"` that we treat in a custom way in the `EditorStyles` component. 

https://github.com/WordPress/wordpress-develop/blob/bf09fe506620678fb82c3b872309edda0ed8ce61/src/wp-includes/block-editor.php#L497-L553

I know that function is in WordPress Core but in Gutenberg we have a hook to override the output of this function `block_editor_settings_all` filter I think.

---

I know that we used this approach in the past for "svg filters" which are also "special styles" that are not just CSS. @ajlende would know more but I think we moved a little bit from it (not entirely sure why though)

matiasbenedetto replied:
Thanks for adding context. The problem I see with this approach is that the backend function is not called when the client-side changes are applied. So, if the user is switching style variations, that is done entirely in the client (as far as I understand), and no Rest API request is sent to the backend. Would this approach be useful in that case?

youknowriad replied:
>  The problem I see with this approach is that the backend function is not called when the client-side changes are applied

This is not a problem, and we already address this in the "GlobalStylesRenderer" like I said [here](https://github.com/WordPress/gutenberg/pull/65019#discussion_r1754975668). We should just make sure the resolution of the fonts is also done within that `useGlobalStylesOutput` hook.

matiasbenedetto replied:
Let's say we add a JSON array of font-families in the backed using `gutenberg_get_block_editor_settings`, we process those families font faces to include the right URLs if that's needed. I get that. If the user switches the content of the fontFamilies theme.json data with data coming from a style variation those families will still have theme relative (`file:/...`) URLs and not the absolute ones which is the reason we are processing the font families in a custom way. I don't see how that can be handled in the code [linked](https://github.com/WordPress/gutenberg/pull/65019#discussion_r1754975668).

youknowriad replied:
you're right in the sense that `useGlobalStylesOutput` is also a block-editor API it seems. I thought personally that it was an "edit-site" API, so we could fetch the theme and do the resolution from there. That said it seems `useGlobalStylesOutput` is a private API for now. I think we can add the "theme path/url" as an argument to address that.

It's unfortunate thought that "resolving fonts" is the only thing from a "theme.json" object that requires outside data (theme) in order to resolve the styles properly.

ramonjd replied:
This touches upon many of the issues we grappled with when dealing with background image relative paths.

Ultimately, we settled on a balance between the server and client, where:

- the server resolves relative paths
- sends them to the client in the `_links` array in the REST response
- the client then has access to both relative and absolute paths for different tasks, e.g., using relative paths for exporting theme.json, and absolute paths for building CSS

If you think something similar would be appropriate here, I'd be happy to help out.

See the original PR: https://github.com/WordPress/gutenberg/pull/61271

Also, there is a _slightly_ related discussion about whether absolute paths to theme assets should be saved in the global styles CPT at all: https://github.com/WordPress/gutenberg/issues/61809

youknowriad replied:
Thanks @ramonjd that is indeed very similar. It seems like we're starting to have multiple needs for the same thing: resolving relative paths during "global styles resolution". It would be good to normalize/document/generalize this a bit so the path is clear if it comes up again.

matiasbenedetto replied:
Thanks for the additional context. I took a look to the `_links` implementation and (at a first glance) it seems to be more complex than needed for this particular dutie (resolve the theme relative path of a font face in the browser). Does the `_links`knowledge of global styles implies that the global styles components know about the rest API design?  It's good the intention to provide a common way to do this but is the backend code added really needed? 

Could injecting the currentTheme into the editor settings using the GlobaStylesRenderer component be a simpler approach that works? I tried that in this commit: https://github.com/WordPress/gutenberg/pull/65019/commits/8ced871bc797e8b4d3819892e01ea7133002938f

https://github.com/WordPress/gutenberg/blob/8ced871bc797e8b4d3819892e01ea7133002938f/packages/edit-site/src/components/global-styles-renderer/index.js#L18-L30

`GlobaStylesRenderer` is inside the `edit-site` package and I imagine that is ok it knows about a theme, is this correct?


Maybe the problem is that we would need to read the theme from the editor settings in the useEditorFontsResolver which is inside `block-editor` package.

https://github.com/WordPress/gutenberg/blob/8ced871bc797e8b4d3819892e01ea7133002938f/packages/block-editor/src/components/use-editor-fonts-resolver/index.js#L13-L25

How bad is this solution?

ramonjd replied:
> resolve the theme relative path of a font face in the browser

What does "resolve" mean in this context? Resolving to get the full URL I presume.

Can you be sure that `currentTheme.stylesheet_uri` is the [base URL](https://github.com/WordPress/gutenberg/pull/65019/files#diff-1db5cd8aa6c0226c4bf0053532d6419d41d6c477b6b9360633c2595a3a49e883R43) of the font assets always?

If so, then it might be enough to do it on the clientside.

One of the considerations with which I had to contend when looking at theme.json image assets is that the file could be in the current theme's directory, or, in the case of a child theme, inherited by the parent theme.

Also there are [filters](https://developer.wordpress.org/reference/hooks/stylesheet_uri/) that allow themes/plugins etc to change the stylesheet URI.

matiasbenedetto replied:
I'm trying to follow the pattern of [this PR](https://github.com/WordPress/gutenberg/pull/61271) using the _links array. It seems to work with the regular theme.json, but I'm unable to access the links for the style variations. How I can pass the variation _settings to the editor settings?

ramonjd replied:
I'm trying to understand the use case. What do we need to do with the resolved font paths? Preload the fonts? Could you provide some testing steps and expectations for style variations? 

Just to confirm, if I run `await wp.data.resolveSelect( 'core' ).__experimentalGetCurrentThemeGlobalStylesVariations()` in the console I see that the variations' fonts are resolved in the `_links` object.

> How I can pass the variation _settings to the editor settings?

Aren't the links already [available in the settings](https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/editor/src/components/provider/use-block-editor-settings.js#L289-L289)?

It's private, so, in the [useEditorFontsResolver](https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/block-editor/src/components/use-editor-fonts-resolver/index.js#L14-L14) `useSelect` you'd have to import the private symbol key (`import { globalStylesLinksDataKey } from '../../store/private-keys';`) and access it like `settings[ globalStylesLinksDataKey ]`.

I just logged the settings out in that hook and I can see the font `_links` object, and when I change variations the links update.

I'm not sure that helps, maybe it leads you in the right direction?


------

Just to clarify and contrast what I was working on with background images, the resolved background image paths are used to build the correct CSS. It's done at this point in the global styles output generation:

https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L308-L314

`tree` contains the embedded `_links` object and path resolution is done "on the fly" when outputting CSS.

matiasbenedetto replied:
> What do we need to do with the resolved font paths? Preload the fonts? Could you provide some testing steps and expectations for style variations?

This is the issue we are trying to fix: https://github.com/WordPress/gutenberg/issues/59965
(load the fonts in the main frame and in the editor iframes when the fonts are defined in variations)

matiasbenedetto replied:
In this [commit](https://github.com/WordPress/gutenberg/pull/65019/commits/146198902f9e68f23681157bcafe4497143f6cdb) I added the JS consumption of the _links. It seems to work for the main editor but not for the editors loaded to display the style variations.



The commit includes a console.log to display the font families and the _links.

https://github.com/WordPress/gutenberg/blob/146198902f9e68f23681157bcafe4497143f6cdb/packages/block-editor/src/components/use-editor-fonts-resolver/index.js#L22-L34

In the screencast you can see how the editors instances created to display the style variations doesn´t get the right list of font families (the ones from the style variation) and the _links array is empty.

[Screencast from 27-09-24 14:04:29.webm](https://github.com/user-attachments/assets/a7c4bc97-78ba-46d2-909b-af872ea9dacf)

Is there any way to get the right list of font families and _links in the editors used to feature a style variation?

ramonjd replied:
So it looks like here you need all the variations' fonts at once for the sidebar. Is that right?

If you look at where the [key value is set in the settings](https://github.com/WordPress/gutenberg/blob/ee98b6ecbefe1d68f6d53d3a52b4550f34f1bade/packages/editor/src/components/provider/use-block-editor-settings.js#L200), you'll notice that the `_links` passed to settings contains only the resolved links for the merged global styles that are currently loaded in the editor. 

That's why the resolved paths for style variations only work once you click on them, because it's it that point they're loaded into the editor.

So you'll probably need to iterate over the output of `await wp.data.resolveSelect( 'core' ).__experimentalGetCurrentThemeGlobalStylesVariations()` as hinted at above since they contain all the variations their resolved paths, and somehow preload the fonts for each variation.

However you can't use these core selectors in the block editor package, that is, in `useEditorFontsResolver()`.

Rather, you might like to tackle from the edit-site package, where the [style variations are rendered](https://github.com/WordPress/gutenberg/blob/146198902f9e68f23681157bcafe4497143f6cdb/packages/edit-site/src/components/global-styles/style-variations-container.js#L124-L124).

For example, export `useEditorFontsResolver()` as a private method and have it accept `_links` and `fontFamilies` as dependencies. I don't know if that will work, but it's the first thing I'd try without having extra context.

justintadlock commented:
> > The two areas where I'm still seeing fallback fonts from the stack are under the Typesets panel:
> 
> @justintadlock, could you provide the theme you are testing?

I wasn't testing with any one specific theme. What I shared above could be seen with any theme that includes web fonts in a style variation. I saw the same with Twenty Twenty-Four.

Do the changes since my last comment address loading the fonts in the iframes in the two sidebar panels? I can certainly test again.

justintadlock commented:
The latest PR is behaving much better than it did when I initially tested. There's still an issue with the initial state in the sidebar panels, but they are updated as you attempt to preview.

Here's a quick video that walks through what's currently happening:

https://github.com/user-attachments/assets/f3ff1949-575c-405c-9a8f-af80cfb9afa9

I zipped a copy of my theme, which has a lot more variety with typography than TT4: [x3p0-ideas.zip](https://github.com/user-attachments/files/17047137/x3p0-ideas.zip)

matiasbenedetto committed:
Merge branch 'trunk' into add/editor-font-face-resolver

matiasbenedetto committed:
Merge branch 'trunk' into add/editor-font-face-resolver

matiasbenedetto committed:
try adding currentTheme to the editor settings

matiasbenedetto committed:
add theme fonts uris

ramonjd commented on lib/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -831,6 +831,50 @@ public static function get_style_variations_from_directory( $directory, $scope =
+	 * @param 
+	 * @return array An array of resolved paths.
+	 */
+	private static function get_resolved_fonts_theme_uris( $theme_json_data ) {
```
I think this could be part of `get_resolved_theme_uris` - it's not that much code, and it could reuse `$placeholder` etc.

<details>

<summary>Example diff</summary>

```diff

diff --git a/lib/class-wp-theme-json-resolver-gutenberg.php b/lib/class-wp-theme-json-resolver-gutenberg.php
index 9c64e09b7e..c128a3c6ea 100644
--- a/lib/class-wp-theme-json-resolver-gutenberg.php
+++ b/lib/class-wp-theme-json-resolver-gutenberg.php
@@ -838,30 +838,30 @@ class WP_Theme_JSON_Resolver_Gutenberg {
 	 *
 	 * @since 6.6.0
 	 *
-	 * @param 
+	 * @param
 	 * @return array An array of resolved paths.
 	 */
 	private static function get_resolved_fonts_theme_uris( $theme_json_data ) {
 		$resolved_theme_uris = array();
 
-		if ( !empty( $theme_json_data['settings']['typography']['fontFamilies'] ) ) {
+		if ( ! empty( $theme_json_data['settings']['typography']['fontFamilies'] ) ) {
 
 			$font_families = ( $theme_json_data['settings']['typography']['fontFamilies']['theme'] ?? array() )
 				+ ( $theme_json_data['settings']['typography']['fontFamilies']['custom'] ?? array() )
 				+ ( $theme_json_data['settings']['typography']['fontFamilies']['default'] ?? array() );
 
 			foreach ( $font_families as $font_family ) {
-				if ( !empty( $font_family['fontFace'] ) ) {
+				if ( ! empty( $font_family['fontFace'] ) ) {
 					foreach ( $font_family['fontFace'] as $font_face ) {
-						if ( !empty( $font_face['src'] ) ) {
-							$sources = is_string( $font_face['src'] ) 
+						if ( ! empty( $font_face['src'] ) ) {
+							$sources = is_string( $font_face['src'] )
 								? array( $font_face['src'] )
 								: $font_face['src'];
 							foreach ( $sources as $source ) {
 								if ( str_starts_with( $source, 'file:' ) ) {
 									$resolved_theme_uris[] = array(
-										'name' => $source,
-										'href' => sanitize_url( get_theme_file_uri( str_replace( 'file:./', '', $source ) ) ),
+										'name'   => $source,
+										'href'   => sanitize_url( get_theme_file_uri( str_replace( 'file:./', '', $source ) ) ),
 										'target' => "typography.fontFamilies.{$font_family['slug']}.fontFace.src",
 									);
 								}
@@ -896,12 +896,6 @@ class WP_Theme_JSON_Resolver_Gutenberg {
 
 		$theme_json_data = $theme_json->get_raw_data();
 
-		// Add font URIs.
-		$resolved_theme_uris = array_merge(
-			$resolved_theme_uris,
-			static::get_resolved_fonts_theme_uris( $theme_json_data )
-		);
-
 		// Using the same file convention when registering web fonts. See: WP_Font_Face_Resolver:: to_theme_file_uri.
 		$placeholder = 'file:./';
 
@@ -952,6 +946,31 @@ class WP_Theme_JSON_Resolver_Gutenberg {
 			}
 		}
 
+		// Add font URIs.
+		if ( ! empty( $theme_json_data['settings']['typography']['fontFamilies'] ) ) {
+			$font_families = array_merge(
+				$theme_json_data['settings']['typography']['fontFamilies']['theme'] ?? array(),
+				$theme_json_data['settings']['typography']['fontFamilies']['custom'] ?? array(),
+				$theme_json_data['settings']['typography']['fontFamilies']['default'] ?? array()
+			);
+			foreach ( $font_families as $font_family ) {
+				if ( ! empty( $font_family['fontFace'] ) ) {
+					foreach ( $font_family['fontFace'] as $font_face ) {
+						$sources = (array) ( $font_face['src'] ?? array() );
+						foreach ( $sources as $source ) {
+							if ( is_string( $source ) && strpos( $source, $placeholder ) === 0 ) {
+								$resolved_theme_uris[] = array(
+									'name'   => $source,
+									'href'   => sanitize_url( get_theme_file_uri( substr( $source, strlen( $placeholder ) ) ) ),
+									'target' => "typography.fontFamilies.{$font_family['slug']}.fontFace.src",
+								);
+							}
+						}
+					}
+				}
+			}
+		}
+
 		return $resolved_theme_uris;
 	}

```

</details>

matiasbenedetto replied:
In don't think that the current design of that kind of function (a very long function processing all the different part of the json data) in both the backend and frontend are the best because they are hard to follow and update because they are very long. Would those function benefit of being more modular?

matiasbenedetto replied:
Still to be in line with the rest of the codebase I moved the functionality inside `get_resolved_theme_uris` here: https://github.com/WordPress/gutenberg/pull/65019/commits/efd95620151a7dba754fb1d1d4407be34ecbdabb

ramonjd commented on lib/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -831,6 +831,50 @@ public static function get_style_variations_from_directory( $directory, $scope =
+		if ( !empty( $theme_json_data['settings']['typography']['fontFamilies'] ) ) {
+
+			$font_families = ( $theme_json_data['settings']['typography']['fontFamilies']['theme'] ?? array() )
+				+ ( $theme_json_data['settings']['typography']['fontFamilies']['custom'] ?? array() )
```
Was the union operator `+` intentional? I think `array_merge()` is more readable and will reindex the new array, avoid potential lost data.

See: https://stitcher.io/blog/array-merge-vs+

Unit test coverage would help test that.

matiasbenedetto replied:
:+1: Using array_merge now

ramonjd commented on packages/block-editor/src/components/use-editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,71 @@
+
+	const { currentTheme = {}, fontFamilies = [] } = useSelect( ( select ) => {
+		return {
+			currentTheme:
+				select( editorStore ).getSettings()?.__experimentalFeatures
+					?.currentTheme,
+			fontFamilies:
+				select( editorStore ).getSettings()?.__experimentalFeatures
+					?.typography?.fontFamilies,
+		};
```
Opportunity to call `getSettings()` once, e.g., 

```js
		const { __experimentalFeatures } =
			select( blockEditorStore ).getSettings();
		return {
			currentTheme: __experimentalFeatures?.currentTheme,
			fontFamilies: __experimentalFeatures?.typography?.fontFamilies,
		};
```

matiasbenedetto replied:
This part of the code was removed. Marking this comment as resolved

ramonjd commented on packages/block-editor/src/components/use-editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,71 @@
+ * Internal dependencies
+ */
+import { getDisplaySrcFromFontFace, loadFontFaceInBrowser } from './utils';
+import { store as editorStore } from '../../store';
```
Let's use `blockEditorStore` to be specific and differentiate from the editor package's store.

```suggestion
import { store as blockEditorStore } from '../../store';
```

matiasbenedetto replied:
This part of the code was removed. Marking this comment as resolved.

matiasbenedetto committed:
use _links in the font resolver

matiasbenedetto committed:
Merge branch 'trunk' into add/editor-font-face-resolver

getdave committed:
Fix PHP linting

getdave committed:
Use correct alias

getdave commented on packages/block-editor/src/components/use-editor-fonts-resolver/index.js:
```diff
@@ -0,0 +1,76 @@
+				return;
+			}
+
+			const src = resolveThemeFontFaceSrc( fontFace.src, _links );
```
```suggestion
			const src = getThemeFontFaceSrc( fontFace.src, _links );
```

Nit: for code clarity purposes we should probably avoid using "resolve" inside an `async` function. I thought this was async.

matiasbenedetto replied:
This part of the code was removed. Marking this comment as resolved.

getdave commented on packages/block-editor/src/components/use-editor-fonts-resolver/utils.js:
```diff
@@ -0,0 +1,65 @@
+	// Removes leading and trailing quotes.
+	output = output.replace( /^["']|["']$/g, '' );
+
+	// Firefox needs the font name to be wrapped in double quotes meanwhile other browsers don't.
+	if ( window.navigator.userAgent.toLowerCase().includes( 'firefox' ) ) {
+		output = `"${ output }"`;
+	}
```
What happens if we just let _all_ browsers have the wrapped version?

matiasbenedetto replied:
This part of the code was removed. Marking this comment as resolved.

getdave commented on packages/block-editor/src/components/use-editor-fonts-resolver/utils.js:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	let output = input.trim();
+	if ( output.includes( ',' ) ) {
+		output = output
+			.split( ',' )
+			// Finds the first item that is not an empty string.
+			.find( ( item ) => item.trim() !== '' )
+			.trim();
+	}
```
Can we describe why we need to do this? I can see _what_ but at a casual glance I'm not sure as to why.

matiasbenedetto replied:
This part of the code was removed. Marking this comment as resolved.

matiasbenedetto committed:
load the fonts as styles

matiasbenedetto committed:
optional chain

matiasbenedetto committed:
remove default not needed

matiasbenedetto committed:
use array_merge instead of +

matiasbenedetto committed:
get the theme font uris inside get_resolved_theme_uris function

matiasbenedetto committed:
Merge branch 'trunk' into add/editor-font-face-resolver

matiasbenedetto committed:
remove unwanted changes

matiasbenedetto committed:
remove unwwanted changes

matiasbenedetto committed:
lint

matiasbenedetto commented:
I'd appreciate another round of reviews. I updated the description of this PR accordingly and added a theme to test the changes quickly.

 I implemented @ramonjd's suggestion to use backend-resolved URLs sent with the rest API to the frontend using the `_links' property and @youknowriad's suggestion to inject the font families as CSS to the editor. In that way, the code is simpler and shorter, and it still does the job but follows common style patterns.

As a follow-up, I would like to improve the rendering of the font variation boxes as requested by @justintadlock, but I think that can be considered outside the scope of this PR.

matiasbenedetto committed:
lint

ramonjd committed:
Adding a unit test to cover the font file resolution in WP_Theme_JSON_Resolver_Gutenberg::get_resolved_theme_uris

ramonjd commented on lib/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -839,6 +839,7 @@ public static function get_style_variations_from_directory( $directory, $scope =
 	 *
 	 * @since 6.6.0
 	 * @since 6.7.0 Added support for resolving block styles.
+	 * @since 6.7.0 Added support for resolving font faces URIs.
```
```suggestion
	 * @since 6.7.0 Added support for resolving block style and font face URIs.
```

ramonjd commented on phpunit/class-wp-theme-json-resolver-test.php:
```diff
@@ -1286,8 +1286,44 @@ public function test_resolve_theme_file_uris() {
 			array(
-				'version' => WP_Theme_JSON_Gutenberg::LATEST_SCHEMA,
-				'styles'  => array(
+				'version'  => WP_Theme_JSON_Gutenberg::LATEST_SCHEMA,
```
I updated the current test to include fonts - hope you don't mind!

matiasbenedetto replied:
:+1: cool thanks

ramonjd commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -913,6 +980,8 @@ export const toStyles = (
 
 	let ruleset = '';
 
+	ruleset += getFontFaceDeclarations( tree );
```
Nice!

ramonjd reviewed (commented):
Thanks for all the work here. 

I tweaked the attached theme slightly so that the previews would use the primary font as well. 

https://github.com/user-attachments/assets/df849b80-0cf2-46d7-a364-ab9476257594

It's working well from my perspective - though some testing from folks with more experience in the font library would be great. 

Happy to test out the backport PR too.

I cheekily added a bit of PHP unit test coverage.

matiasbenedetto committed:
update comment

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

getdave reviewed (commented):
I used your Theme for testing (thanks so much for taking the time to make testing easier) and the fonts loaded on demand as I switched between the Style Variations.

The one thing I noticed is that the fonts are requested again, even if they have previously been loaded.

In the screencapture below you'll see the requests to fonts going out again the second time round.

I don't think it represents a huge problem though, as even on slow network, caching seems to kick in and the fonts are displayed immediately despite the unwanted network request.


https://github.com/user-attachments/assets/0253e0e4-2f4c-415c-9522-ef6b9ac9807b

getdave commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -884,6 +885,72 @@ export const toCustomProperties = ( tree, blockSelectors ) => {
+
+	const themeFileURIs = tree?._links?.[ 'wp:theme-file' ] ?? [];
+
+	// Iterate over main origins (theme, custom, default)
```
Nit: It could be helpful to include a comment here explaining what this block of code does.

Presumably it's generating @font-face declarations so an example of the generated output would help folks to conceptualise this.

matiasbenedetto replied:
:+1: Added comment here: https://github.com/WordPress/gutenberg/pull/65019/commits/b69ae26276c3db6519fc790d1415500373363bfc

getdave commented:
@kevin940726 @ndiego @colorful-tones Shall we pursue this for 6.7?

ramonjd commented:
> Shall we pursue this for 6.7?

For visibility, there's an alleged, but yet unconfirmed performance regression in the way Theme JSON resolver is resolving theme JSON relative paths. 

Hopefully it will be confirmed, and even better, addressed this week if it is indeed found to be correct.

https://github.com/WordPress/performance/issues/1572#issuecomment-2412996324

If it is real, I think adding more checks might exacerbate things.

colorful-tones commented:
> Hopefully it will be confirmed, and even better, addressed this week if it is indeed found to be correct.

Based on the progress here I would say we can keep it on the radar for 6.7, but not a solid 'yes'. 😃

matiasbenedetto committed:
add function comment

matiasbenedetto commented:
> The one thing I noticed is that the fonts are requested again, even if they have previously been loaded.
> In the screencapture below you'll see the requests to fonts going out again the second time round.
> I don't think it represents a huge problem though, as even on slow network, caching seems to kick in and the fonts are displayed immediately despite the unwanted network request.

@getdave thanks for testing!
In the screencast you have the 'disable cache' option of the browser debug bar enabled. If you uncheck that (default browser behavior when the debug bar is closed) the browser is able to cache the fonts and they are not requested again if it was already loaded. See the comparison in the screencast with the 'disable cache' option checked and unchecked:

[Screencast from 15-10-24 14:33:47.webm](https://github.com/user-attachments/assets/1020623d-ba06-4167-bc81-90ce191e19a2)

matiasbenedetto commented:
**Trac ticket created**: https://core.trac.wordpress.org/ticket/62231
**Wordpress-develop PR created**: https://github.com/WordPress/wordpress-develop/pull/7570

ramonjd commented:
> Hopefully it will be confirmed, and even better, addressed this week if it is indeed found to be correct.

Update: @aaronrobertshaw and I spent a big chunk of the day testing and couldn't confirm the regression, so I think we're good to go.

Moreover, the CI performance [tests are happy](https://github.com/WordPress/wordpress-develop/actions/runs/11352787760/job/31576330178?pr=7570) as far as I can tell.

I think the file resolution probably needs to be cached at some point. I have it on my list to look at.

Thank you!

getdave commented:
> In the screencast you have the 'disable cache' option of the browser debug bar enabled

Haha so I did. Good catch. I did think it was strange!

matiasbenedetto commented:
I submitted a simpler alternative to this PR to fix the problem directly in core: https://github.com/WordPress/wordpress-develop/pull/7581

carolinan commented:
I'm sorry I have not find the time to test this.
Truthfully I don't understand the code in this PR, so I would not be able to do a code review.

But I don't want you to feel like you are forced to come up with another option (7581) only because of lack of reviews.

getdave commented:
> I submitted a simpler alternative to this PR to fix the problem directly in core: [WordPress/wordpress-develop#7581](https://github.com/WordPress/wordpress-develop/pull/7581)

@matiasbenedetto I noticed this. Is there a reason why you decided to raise directly in Core? Is is a more suitable place for the fix? Does adding to Gutenberg just cause more overhead? Is it a problem of reviews? 

If you're lacking reviews and this then I'm happy to flag for the release leads to take a look at.

matiasbenedetto commented:
> But I don't want you to feel like you are forced to come up with another option (7581) only because of lack of reviews.

> Is there a reason why you decided to raise directly in Core? Is is a more suitable place for the fix? Does adding to Gutenberg just cause more overhead? Is it a problem of reviews?

@carolinan @getdave, thanks for your consideration! I submitted https://github.com/WordPress/wordpress-develop/pull/7581 because I think it's a simpler approach, easier to debug and extend, and, for that reason, better.  It's not about reviews.

kevin940726 commented:
Sorry, I came to this late. Are we still aiming this for 6.7? Should we focus on landing https://github.com/WordPress/wordpress-develop/pull/7581 and close this? Sorry if I'm misunderstanding something! 🙇

andrewserong commented:
> Sorry, I came to this late. Are we still aiming this for 6.7? Should we focus on landing https://github.com/WordPress/wordpress-develop/pull/7581 and close this? Sorry if I'm misunderstanding something! 🙇

It looks like 7581 has been committed (https://github.com/WordPress/wordpress-develop/pull/7581#issuecomment-2425285546), so I've removed the Backport label for 6.7 and moved this PR to punted to 6.8 in the project board, as my understanding is that this PR is no longer required for 6.7.

Feel free to update if that's not correct @matiasbenedetto! I wasn't sure if you still wanted to work on this PR, so have otherwise left it open for now.

matiasbenedetto commented:
Thanks everyone for the feedback.
I think we can close this. https://github.com/WordPress/wordpress-develop/pull/7581 should fix the original issue in core.
